pytorch_lightning/metrics/classification/utils.py

```diff
@@ -0,0 +1,390 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Tuple, Optional
+
+import torch
+
+from pytorch_lightning.metrics.utils import to_onehot, select_topk
+
+
+def _check_classification_inputs(
+    preds: torch.Tensor,
+    target: torch.Tensor,
+    threshold: float,
+    num_classes: Optional[int] = None,
+    is_multiclass: bool = False,
```

Comment:
Is this argument necessary if we settle on num_classes = 1 means binary and num_classes!=1 is multiclass?

Reply:
Sooo right now the "division of labor" is that `num_classes` is there to determine the shape for one-hot transformation (if needed), and consistency/error checking - if you specified `num_classes` in init, this would make sure your data indeed has that many classes, as I think it is more graceful to catch the error here than when you are updating the state tensors.

And `is_multiclass` is there to determine if the inputs need to be transformed to other types. If we make `num_classes` do all the work, you may loose some error checking power: for example if you have, for some reason, ML data but with binary integer preds (this is one of the inputs in tests), then to signal that it is multi-label you would, without `is_multiclass`, need to pass `num_classes=1`. This is both a bit strange (number of classes is actually `C`), and prevents you from checking that number of classes is whatever was set at init.

Reply:
I've tried it wit `num_classes` doing everything before, but I've decided to add this extra parameter for the reasons above (and maybe I forgot something as well). Also, I think it's easier to explain this to the users, rather than adding some convoluted discussion on how `num_classes` can be used.

Reply:
I have to think a bit more about this, but here is a counter-argument:
Sklearn does not have this kind of argument, and therefore I believe it should absolutely be possible to differentiate between the different cases without. As a user, it feels strange that if I specify `num_classes=5` for example I also have to specify that `is_multiclass=True`. It seems redundant.

Reply:
There's no situation where you have to specify `num_classes=5` and `is_multiclass=True`

Reply:
`is_multiclass=True` is only unsed to convert binary and multi-label data into 2-class (multi-dimensional) multi-class

Reply:
Sklearn does not have this because, for example in recall, it has this extra average "binary", which users have to specify if they want to treat their data as binary. So they also can't differentiate between the classes automatically, but instead offload this to different parameters for each metric. I think my way is much better/clearer, as this is not about the the "averaging method", but about the nature of your data.

Reply:
Not to mention that sklearn does not deal with multi-dim multi-class data at all